Sources/SwiftParser/Statements.swift

```diff
@@ -877,11 +918,66 @@ extension Parser.Lookahead {
         // For example, could be the function call "discard()".
         return false
       }
+
+    case .then:
+      return isStartOfThenStatement(preferExpr: preferExpr)
+
     case nil:
       return false
     }
   }
 
+  /// Whether we're currently at a `then` token that should be parsed as a
+  /// `then` statement.
+  mutating func isStartOfThenStatement(preferExpr: Bool) -> Bool {
+    // If the feature is disabled, don't parse it.
+    guard self.experimentalFeatures.contains(.thenStatements) else {
+      return false
+    }
+    guard self.at(.keyword(.then)) else {
+      return false
+    }
+
+    // If we want to prefer an expr, and aren't at the start of a newline, then
+    // don't parse a ThenStmt.
+    if preferExpr && !self.atStartOfLine {
+      return false
+    }
+
+    let next = peek()
+
+    // If 'then' is followed by a binary or postfix operator, prefer to parse as
+    // an expr.
+    if BinaryOperatorLike(lexeme: next) != nil || PostfixOperatorLike(lexeme: next) != nil {
+      return false
+    }
+
+    switch next.rawTokenKind {
+    case .keyword:
+      // Treat 'is' and 'as' like the binary operator case, and parse as an
+      // expr.
+      switch Keyword(next.tokenText) {
+      case .as, .is:
+        return false
+      default:
+        break
+      }
+
+    case .leftBrace:
+      // This is a trailing closure.
+      return false
```

Comment:
I also commented on a test case below but: We could also be `then`ing (or however you call the action you do with a 'then' statement) a closure. Is that just not supported, or you need to wrap the closure in parentheses?

Reply:
Let's continue this discussion on that original comment